botocore/useragent.py

```diff
@@ -0,0 +1,421 @@
+_USERAGENT_ALLOWED_OS_NAMES = [
+    'windows',
+    'linux',
+    'macos',
+    'android',
+    'ios',
+    'watchos',
+    'tvos',
+    'other',
+]
+_USERAGENT_PLATFORM_NAME_MAPPINGS = {'darwin': 'macos'}
+# The name by which botocore is identified in the User-Agent header. While most
+# AWS SDKs follow a naming pattern of "aws-sdk-*", botocore and boto3 continue
+# using their existing values. Uses uppercase "B" with all other characters
+# lowercase.
+_USERAGENT_SDK_NAME = 'Botocore'
+
+
+def sanitize_user_agent_string_component(raw_str, allow_hash):
+    """Replaces all not allowed characters in the string with a dash ("-").
+
+    Allowed characters are ASCII alphanumerics and ``!$%&'*+-.^_`|~``. If
+    ``allow_hash`` is ``True``, "#"``" is also allowed.
+
+    :type allow_hash: bool
+    :param allow_hash: Whether "#" is considered an allowed character.
+    """
+    return ''.join(
+        c
+        if c in _USERAGENT_ALLOWED_CHARACTERS or (allow_hash and c == '#')
+        else '-'
+        for c in raw_str
+    )
+
+
+class UserAgentComponent(NamedTuple):
+    """
+    Component of a Botocore User-Agent header string in the standard format.
+
+    Each component consists of a prefix, a name, and a value. In the string
+    representation these are combined in the format ``prefix/name#value``.
+    """
+
+    prefix: str
+    name: str
+    value: Optional[str] = None
+
+    def to_string(self):
+        """Create string like 'prefix/name#value' from a UserAgentComponent."""
+        clean_prefix = sanitize_user_agent_string_component(
+            self.prefix, allow_hash=True
+        )
+        clean_name = sanitize_user_agent_string_component(
+            self.name, allow_hash=False
+        )
+        if self.value is None or self.value == '':
+            return f'{clean_prefix}/{clean_name}'
+        clean_value = sanitize_user_agent_string_component(
+            self.value, allow_hash=True
+        )
+        return f'{clean_prefix}/{clean_name}#{clean_value}'
+
+
+class RawStringUserAgentComponent:
+    """
+    UserAgentComponent interface wrapper around ``str``.
+
+    Use for User-Agent header components that are not constructed from
+    prefix+name+value but instead are provided as strings. No sanitization is
+    performed.
+    """
+
+    def __init__(self, value):
+        self._value = value
+
+    def to_string(self):
+        return self._value
+
+
+class UserAgentString:
+    """
+    Generator for AWS SDK User-Agent header strings.
+
+    The User-Agent header format contains information from session, client, and
+    request context. ``UserAgentString`` provides methods for collecting the
+    information and ``to_string`` for assembling it into the standardized
+    string format.
+
+    Example usage:
+
+        ua_session = UserAgentString.from_environment()
+        ua_session.set_session_config(...)
+        ua_client = ua_session.with_client_config(Config(...))
+        ua_string = ua_request.to_string()
```

Comment:
There seems to be inconsistent naming here. Both should be `ua_client` or `ua_request`.

Reply:
I don't disagree with the naming being somewhat confusing. But I don't think your suggestions would clarify things. In fact your suggestions might be a sign that I successfully confused you!

* `ua_session` is a `UserAgentString` object with session-level information only.
* `ua_client` is a `UserAgentString` object with session-level and client-level information.
* `ua_string` is a `str`.